### PR TITLE
feat: [PL-43385]: NG Manager Move Project API

### PR DESCRIPTION
**Pull request by adiyaar24**
## Describe your changes
This PR Implements move project API on NG Manager, which is Internal and Hidden .

## Checklist
- [x] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/56588)

### CODE REVIEW

**Comment by adiyaar24** [comment]
trigger uts

File: 120-ng-manager/src/main/java/io/harness/ng/core/remote/ProjectResource.java [hunk 1]
```diff
@@ -316,6 +322,31 @@ public class ProjectResource {
             .build());
   }
 
+  @PUT
+  @Path("{identifier}/move")
+  @InternalApi
```
Comment (pbarapatre10) [inline 120-ng-manager/src/main/java/io/harness/ng/core/remote/ProjectResource.java:327]: Hidden is sufficient, if we mark internal we woyuld need service to service token to call this API even from POSTMAN, thanks

File: 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java [hunk 2]
```diff
@@ -495,6 +497,54 @@ public Project update(String accountIdentifier, @OrgIdentifier String orgIdentif
         USER);
   }
 
+  public boolean moveProject(String accountIdentifier, @OrgIdentifier String orgIdentifier,
+      @ProjectIdentifier String identifier, @OrgIdentifier String destinationOrgIdentifier) {
```
Comment (pbarapatre10) [inline 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java:501]: Why the `@OrgIdentifier` is needed for destinationOrg? thanks

File: 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java [hunk 3]
```diff
@@ -495,6 +497,54 @@ public Project update(String accountIdentifier, @OrgIdentifier String orgIdentif
         USER);
   }
 
+  public boolean moveProject(String accountIdentifier, @OrgIdentifier String orgIdentifier,
```
Comment (pbarapatre10) [inline 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java:500]: This API should ideally take `ScopeInfo` as input and not the current `orgIdentifier`, thanks

File: 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java [hunk 4]
```diff
@@ -495,6 +497,54 @@ public Project update(String accountIdentifier, @OrgIdentifier String orgIdentif
...
+      @ProjectIdentifier String identifier, @OrgIdentifier String destinationOrgIdentifier) {
+    Optional<Organization> organizationOptional = organizationService.get(accountIdentifier, destinationOrgIdentifier);
```
Comment (pbarapatre10) [inline 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java:502]: Please update to use organizationService.get with `scopeInfo`

File: 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java [hunk 5]
```diff
@@ -495,6 +497,54 @@ public Project update(String accountIdentifier, @OrgIdentifier String orgIdentif
...
+    if (organizationOptional.isEmpty()) {
+      throw new EntityNotFoundException(String.format("Organization with identifier [%s] not found", orgIdentifier));
+    }
+    Optional<Project> duplicateProjectCheck = get(accountIdentifier, destinationOrgIdentifier, identifier);
```
Comment (pbarapatre10) [inline 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java:506]: Same as here, lets try to use internal get with `scopeInfo`, thanks

File: 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java [hunk 6]
```diff
@@ -495,6 +497,54 @@ public Project update(String accountIdentifier, @OrgIdentifier String orgIdentif
...
+    if (duplicateProjectCheck.isPresent()) {
+      throw new DuplicateFieldException(
+          String.format("A project with identifier [%s] and orgIdentifier [%s] is already present",
+              duplicateProjectCheck.get().getIdentifier(), destinationOrgIdentifier),
+          USER_SRE);
```
Comment (pbarapatre10) [inline 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java:511]: change `USER_SRE` to `USER` thanks

File: 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java [hunk 7]
```diff
@@ -495,6 +497,54 @@ public Project update(String accountIdentifier, @OrgIdentifier String orgIdentif
...
+    }
+    Optional<Project> optionalProject = get(accountIdentifier, orgIdentifier, identifier);
```
Comment (pbarapatre10) [inline 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java:513]: same as above, use scopeInfo, thanks

File: 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java [hunk 8]
```diff
@@ -495,6 +497,54 @@ public Project update(String accountIdentifier, @OrgIdentifier String orgIdentif
...
+
+      project.setModules(project.getModules());
+      validate(project);
+      return Failsafe.with(DEFAULT_RETRY_POLICY).get(() -> transactionTemplate.execute(status -> {
+        Project updatedProject = projectRepository.save(project);
+        addToScopeInfoCache(updatedProject);
+        setupProject(Scope.of(accountIdentifier, destinationOrgIdentifier, project.getIdentifier()));
+        log.info(String.format(
+            "Project with identifier [%s] and source orgIdentifier [%s] was successfully migrated to [%s] orgIdentifier",
+            identifier, orgIdentifier, destinationOrgIdentifier));
```
Comment (pbarapatre10) [inline 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java:526]: Lets also log source 'scopeUniqueId' and destination 'scopeUniqueId' is available, thanks

File: 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java (hunk 5 above)
Comment (pbarapatre10) [reply 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java:495]: For destination scopeInfo might not work, so ignore if not applicable, thanks

File: 120-ng-manager/src/main/java/io/harness/ng/core/remote/ProjectResource.java [hunk 9]
```diff
@@ -316,6 +322,31 @@ public class ProjectResource {
...
+      @Parameter(description = PROJECT_PARAM_MESSAGE) @NotNull @PathParam(
+          NGCommonEntityConstants.IDENTIFIER_KEY) @ResourceIdentifier String identifier,
+      @Parameter(description = ACCOUNT_PARAM_MESSAGE) @NotNull @QueryParam(
+          NGCommonEntityConstants.ACCOUNT_KEY) @AccountIdentifier String accountIdentifier,
+      @Parameter(
+          description = "Organization identifier for the Project. If left empty, Default Organization is assumed")
+      @QueryParam(NGCommonEntityConstants.ORG_KEY) @DefaultValue(
+          DEFAULT_ORG_IDENTIFIER) @OrgIdentifier String orgIdentifier,
+      @RequestBody(required = true,
+          description = "This is the updated org identifier") @NotNull @Valid MoveProjectRequest moveProjectRequest) {
```
Comment (pbarapatre10) [inline 120-ng-manager/src/main/java/io/harness/ng/core/remote/ProjectResource.java:342]: Add `@Context ScopeInfo scopeinfo` as a resource param and that will be propagated to the service, thanks

File: 960-ng-core-beans/src/main/java/io/harness/ng/core/dto/MoveProjectRequest.java [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class MoveProjectRequest {
```
Comment (pbarapatre10) [inline 960-ng-core-beans/src/main/java/io/harness/ng/core/dto/MoveProjectRequest.java:20]: nit: add `OwnedBy(PL)`, thanks

File: 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java (hunk 5 above)
Comment (adiyaar24) [reply 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java:495]: But I am checking at the destination org Identifier, so thereby I need to do this way

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56588)* status: 6 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java` line 523 at r3](https://reviewable.io/reviews/harness/harness-core/56588#-NlXX3Xc9LuaMeQyC6cO:-NlXX3Xc9LuaMeQyC6cP:b-4ym7uu) ([raw file](https://github.com/harness/harness-core/blob/23f88dba7b7920b55c066b00242158644e99da61/120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java#L523)):*
> ```Java
>         Project updatedProject = projectRepository.save(project);
>         addToScopeInfoCache(updatedProject);
>         setupProject(Scope.of(accountIdentifier, destinationOrgIdentifier, project.getIdentifier()));
> ```

What this method will do internally?
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java` line 525 at r3](https://reviewable.io/reviews/harness/harness-core/56588#-NlXX0mhAuPqeYL5ff2j:-NlXX0mhAuPqeYL5ff2k:b-x7cf0u) ([raw file](https://github.com/harness/harness-core/blob/23f88dba7b7920b55c066b00242158644e99da61/120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java#L525)):*
> ```Java
>         setupProject(Scope.of(accountIdentifier, destinationOrgIdentifier, project.getIdentifier()));
>         log.info(String.format(
>             "Project with identifier [%s] and source orgIdentifier [%s] was successfully migrated to [%s] orgIdentifier",
> ```

Moved instead of migrated
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java` line 527 at r3](https://reviewable.io/reviews/harness/harness-core/56588#-NlXWv3k29qAA8dUhlb-:-NlXWv3k29qAA8dUhlb0:b43pyng) ([raw file](https://github.com/harness/harness-core/blob/23f88dba7b7920b55c066b00242158644e99da61/120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java#L527)):*
> ```Java
>             "Project with identifier [%s] and source orgIdentifier [%s] was successfully migrated to [%s] orgIdentifier",
>             identifier, orgIdentifier, destinationOrgIdentifier));
>         outboxService.save(new ProjectUpdateEvent(
> ```

Should this be projectMoveEvent?
___
*[`960-ng-core-beans/src/main/java/io/harness/ng/core/dto/MoveProjectRequest.java` line 20 at r2](https://reviewable.io/reviews/harness/harness-core/56588#gh-1424869763:-NlXXRxw0rpDPDX-rZmA:bdp0dqv) ([raw file](https://github.com/harness/harness-core/blob/a2ebf948ffd6a4d22a006b0bb509fbc8787b2385/960-ng-core-beans/src/main/java/io/harness/ng/core/dto/MoveProjectRequest.java#L20)):*
<details><summary><i>Previously, pbarapatre10 (Prateek Barapatre) wrote…</i></summary><blockquote>

nit: add `OwnedBy(PL)`, thanks
</blockquote></details>

Shouldn't this also contain accountIdentifier, sourceOrgIdentifier?

**Review by ankit-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56588)* status: 6 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java` line 491 at r3](https://reviewable.io/reviews/harness/harness-core/56588#-NlXY9so2ZouGBSG3XMg:-NlXY9so2ZouGBSG3XMh:bzepnej) ([raw file](https://github.com/harness/harness-core/blob/23f88dba7b7920b55c066b00242158644e99da61/120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java#L491)):*
> ```Java
>   public boolean moveProject(String accountIdentifier, @OrgIdentifier String orgIdentifier,
>       @ProjectIdentifier String identifier, @OrgIdentifier String destinationOrgIdentifier) {
>     Optional<Organization> organizationOptional = organizationService.get(accountIdentifier, destinationOrgIdentifier);
> ```

destinationOrgOptional

**Review by ankit-harness — DISMISSED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/56588)* status: 6 file(s) to review. 1 change requests. 2. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java` line 527 at r3](https://reviewable.io/reviews/harness/harness-core/56588#-NlXYryj6QeMPzQuibHI:-NlXYryj6QeMPzQuibHJ:b-4ok00j) ([raw file](https://github.com/harness/harness-core/blob/23f88dba7b7920b55c066b00242158644e99da61/120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java#L527)):*
> ```Java
>             "Project with identifier [%s] and source orgIdentifier [%s] was successfully migrated to [%s] orgIdentifier",
>             identifier, orgIdentifier, destinationOrgIdentifier));
>         outboxService.save(new ProjectUpdateEvent(
> ```

How will this event show up in Audit YAML diff? Update to orgIdentifier or update to parentUniqueId field? Do we expose uniqueId and parentUniqueId in YAMLs?

File: 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java [hunk 11]
```diff
@@ -523,6 +529,9 @@ public boolean moveProject(String accountIdentifier, ScopeInfo scopeInfo, @OrgId
         Project updatedProject = projectRepository.save(project);
         addToScopeInfoCache(updatedProject);
         setupProject(Scope.of(accountIdentifier, destinationOrgIdentifier, project.getIdentifier()));
+
+        harnessSMManager.createHarnessSecretManager(accountIdentifier, destinationOrgIdentifier, identifier);
+
```
Comment (pbarapatre10) [inline 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java:534]: nit: should we trigger `ProjectCreate` event and let it take care of this? I'm fine to have it here for now, thanks

**Review by pbarapatre10 — DISMISSED** [review]

File: 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java (hunk 11 above)
Comment (adiyaar24) [reply 120-ng-manager/src/main/java/io/harness/ng/core/impl/ProjectServiceImpl.java:534]: Created a handleProjectMove , will discuss in call

**Comment by adiyaar24** [comment]
trigger uts

**Review by pbarapatre10 — APPROVED** [review]

**Pull request merged**